e2e_playwright/multipage_apps_v2/mpa_v2_basics_test.py

```diff
@@ -0,0 +1,336 @@
+    """Test that we show a not found dialog if the page doesn't exist."""
+    page.goto(f"http://localhost:{app_port}/not_a_page")
+    wait_for_app_loaded(page)
+
+    expect(page.locator('[role="dialog"]')).to_contain_text("Page not found")
+
+
+def test_handles_expand_collapse_of_mpa_nav_correctly(
+    themed_app: Page, assert_snapshot: ImageCompareFunction
+):
+    """Test that we handle expand/collapse of MPA nav correctly."""
+
+    check_field(themed_app, add_sidebar_elements=True)
+    wait_for_app_run(themed_app)
+
+    view_button = themed_app.get_by_test_id("stSidebarNavViewButton")
+
+    expect(view_button).to_be_visible()
+
+    # Expand the nav
+    view_button.click(force=True)
+    # We apply a quick timeout here so that the UI has some time to
+    # adjust for the screenshot after the click
+    themed_app.wait_for_timeout(250)
+    assert_snapshot(
+        themed_app.get_by_test_id("stSidebarNav"), name="mpa-sidebar_nav_expanded"
+    )
+
+    # Collapse the nav
+    view_button.click(force=True)
+    themed_app.wait_for_timeout(250)
+    assert_snapshot(
+        themed_app.get_by_test_id("stSidebarNav"), name="mpa-sidebar_nav_collapsed"
+    )
+
+    # Expand the nav again
+    view_button.click(force=True)
+    themed_app.wait_for_timeout(250)
+    assert_snapshot(
+        themed_app.get_by_test_id("stSidebarNav"), name="mpa-sidebar_nav_expanded"
+    )
+
+
+def test_switch_page_by_path(app: Page):
+    """Test that we can switch between pages by triggering st.switch_page with a path."""
+
+    app.get_by_test_id("baseButton-secondary").filter(has_text="page 5").click()
+    wait_for_app_run(app)
+
+    expect(page_heading(app)).to_contain_text("Page 5")
+
+
+def test_switch_page_by_st_page(app: Page):
+    """Test that we can switch between pages by triggering st.switch_page with st.Page."""
+
+    app.get_by_test_id("baseButton-secondary").filter(has_text="page 9").click()
+    wait_for_app_run(app)
+
+    expect(page_heading(app)).to_contain_text("Page 9")
+
+
+def test_switch_page_removes_query_params(page: Page, app_port: int):
+    """Test that query params are removed when navigating via st.switch_page"""
+
+    # Start at main page with query params
+    page.goto(f"http://localhost:{app_port}/?foo=bar")
+    wait_for_app_loaded(page)
+
+    # Trigger st.switch_page
+    page.get_by_test_id("baseButton-secondary").filter(has_text="page 5").click()
+    wait_for_app_loaded(page)
+    # Check that query params don't persist
+    expect(page).to_have_url(f"http://localhost:{app_port}/page_5")
+
+
+def test_removes_query_params_when_swapping_pages(page: Page, app_port: int):
+    """Test that query params are removed when swapping pages"""
+
+    page.goto(f"http://localhost:{app_port}/page_7?foo=bar")
+    wait_for_app_loaded(page)
+
+    get_page_link(page, "page 4").click()
+    wait_for_app_loaded(page)
+    expect(page).to_have_url(f"http://localhost:{app_port}/page_4")
```

Comment:
nit: might be worth more loudly calling out the difference between these two tests as it took a second for me to realize how they're not duplicates